demos/mqtt/mqtt_demo_lightweight/mqtt_demo_lightweight.c

```diff
@@ -0,0 +1,1045 @@
+    {
+        LogInfo( ( "Incoming Publish Topic Name: %.*s matches subscribed topic.\n",
+                   pPubInfo->topicNameLength,
+                   pPubInfo->pTopicName ) );
+        LogInfo( ( "Incoming Publish Message : %.*s\n",
+                   pPubInfo->payloadLength,
+                   pPubInfo->pPayload ) );
+    }
+    else
+    {
+        LogError( ( "Incoming Publish Topic Name: %.*s does not match subscribed topic. \n",
+                    pPubInfo->topicNameLength,
+                    pPubInfo->pTopicName ) );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Receive and validate MQTT packet from the broker, determine the type
+ * of the packet and processes the packet based on the type.
+ *
+ * @param[in] tcpSocket is a TCP socket that is connected to an MQTT broker to which
+ * an MQTT connection has been established.
+ *
+ * @param[in] pFixedBuffer Pointer to a structure containing fixed buffer and length.
+ *
+ */
+static void mqttProcessIncomingPacket( int tcpSocket,
+                                       MQTTFixedBuffer_t * pFixedBuffer )
+{
+    MQTTStatus_t result;
+    MQTTPacketInfo_t incomingPacket;
+    MQTTPublishInfo_t publishInfo;
+    uint16_t packetId;
+    int status;
+    bool sessionPresent = false;
+    uint16_t receiveAttempts = 0;
+
+    /***
+     * For readability, error handling in this function is restricted to the use of
+     * asserts().
+     ***/
+
+    memset( ( void * ) &incomingPacket, 0x00, sizeof( MQTTPacketInfo_t ) );
+
+    /* Determine incoming packet type and remaining length. */
+    do
+    {
+        /* Retry till data is available */
+        result = MQTT_GetIncomingPacketTypeAndLength( transportRecv, tcpSocket, &incomingPacket );
+
+        /* If it is success of any other failure than transportTimeout, exit immediately */
+        if( result != MQTTNoDataAvailable )
+        {
+            break;
+        }
+
+        receiveAttempts++;
+    } while ( ( result != MQTTSuccess ) && ( receiveAttempts < MQTT_MAX_RECV_ATTEMPTS ) );
+
+    assert( result == MQTTSuccess );
+    assert( incomingPacket.remainingLength <= pFixedBuffer->size );
+
+    /* Current implementation expects an incoming Publish and three different
+     * responses ( SUBACK, PINGRESP and UNSUBACK ). */
+
+    /* Receive the remaining bytes. */
+    status = recv( tcpSocket, ( void * ) pFixedBuffer->pBuffer, incomingPacket.remainingLength, 0 );
+    assert( status == ( int ) incomingPacket.remainingLength );
+
+    incomingPacket.pRemainingData = pFixedBuffer->pBuffer;
+
+    if( ( incomingPacket.type & 0xf0 ) == MQTT_PACKET_TYPE_PUBLISH )
+    {
+        result = MQTT_DeserializePublish( &incomingPacket, &packetId, &publishInfo );
+        assert( result == MQTTSuccess );
+
+        /* Process incoming Publish message. */
+        mqttProcessIncomingPublish( &publishInfo, packetId );
```

Comment:
This is QoS 0, so `MQTT_DeserializePublish()` won't set `packetId`, meaning it will be uninitialized here. Althought `mqttProcessIncomingPublish()` doesn't actually use `packetId`, it'd be better to initialize it to 0 in line 809.